libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationRouteProcessor.kt

```diff
@@ -0,0 +1,141 @@
+package com.mapbox.services.android.navigation.v5.internal.navigation
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.geojson.Geometry
+import com.mapbox.geojson.Point
+import com.mapbox.navigator.NavigationStatus
+import com.mapbox.navigator.RouteState
+import com.mapbox.services.android.navigation.v5.routeprogress.CurrentLegAnnotation
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgressStateMap
+
+internal class NavigationRouteProcessor {
+    private val ONE_INDEX = 1
+    private val ONE_SECOND_IN_MILLISECONDS = 1000.0
+    private val FIRST_BANNER_INSTRUCTION = 0
+
+    private val progressStateMap = RouteProgressStateMap()
+    private var previousRouteProgress: RouteProgress? = null
+    private var previousStatus: NavigationStatus? = null
+    private var route: DirectionsRoute? = null
+    private var currentLeg: RouteLeg? = null
+    private var currentStep: LegStep? = null
+    private var currentStepPoints: List<Point>? = null
+    private var upcomingStepPoints: List<Point>? = null
+    private var currentLegAnnotation: CurrentLegAnnotation? = null
+    private var routeGeometryWithBuffer: Geometry? = null
+
+    fun buildNewRouteProgress(navigator: MapboxNavigator, status: NavigationStatus, route: DirectionsRoute): RouteProgress? {
```

Comment:
Shouldn't be non-nullable? That was the case before the changes.

Reply:
Yes, it should, but the  original code did not guaranty that this variable is not null. The route can be null.  And if the route is null then buildNewRouteProgress() can return a null value.